File: Minimum_Number_of_Taps_to_Open_to_Water_a_Garden.cpp
class Solution {
public:
    int minTaps(int n, vector<int>& ranges) {
        vector<int> maxRange(n + 1, 0);
        
        for (int i = 0; i < ranges.size(); i++) {
            int left = max(0, i - ranges[i]);
            int right = min(n, i + ranges[i]);
            maxRange[left] = max(maxRange[left], right);
        }
        
        int end = 0;
        int farthest = 0;
        int taps = 0;
        int i = 0;
        
        while (end < n) {
            while (i <= end) {
                farthest = max(farthest, maxRange[i]);
                i++;
            }
            
            if (farthest <= end) {
                return -1;  // Unable to cover the entire garden
            }
            
            end = farthest;
            taps++;
        }
        
        return taps;        
    }
};

File: Minimum_Number_of_Taps_to_Open_to_Water_a_Garden.py
class Solution:
    def minTaps(self, n: int, ranges: List[int]) -> int:
        max_range = [0] * (n + 1)

        for i in range(len(ranges)):
            left = max(0, i - ranges[i])
            right = min(n, i + ranges[i])
            max_range[left] = max(max_range[left], right)
        
        end = farthest = taps = 0
        i = 0

        while end < n:
            while i <= end:
                farthest = max(farthest, max_range[i])
                i += 1
            
            if farthest <= end:
                return -1
            
            end = farthest
            taps += 1
        
        return taps
